Bucket diagnostics

AlexisRenchon opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
This PR adds diagnostics to ClimaLand, as a new module in /src.
This module is built on ClimaDiagnostics.jl, which is also used by
ClimaAtmos. It contains default diagnostics methods for different
ClimaLand models (for now, BucketModel and SoilCanopyModel), those
diagnostics contains metadata such as where diagnostics variables are
stored in those models, what is there names (short, long, standard)
and physical units, with additional comments. It creates a Dict of those
variables, and also default Scheduling methods, such as hourly average
or monthly maximum. Users can define their own diagnostic variables or
scheduling, and developpers can add defaults. More information are
available in the documentation.

AlexisRenchon commented:
@Sbozzolo 

Do you know why ClimaLandSimulations CI fail? 
The error is 
```
Run julia --project= -e 'using Pkg; Pkg.develop(path="$(pwd())"); Pkg.develop(path="$(pwd())/lib/ClimaLandSimulations")'
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package ClimaDiagnostics [1ecacbb8]:
 ClimaDiagnostics [1ecacbb8] log:
 ├─possible versions are: 0.0.1-0.0.5 or uninstalled
 └─restricted to versions 0.2 by ClimaLand [08f[4](https://github.com/CliMA/ClimaLand.jl/actions/runs/9321596680/job/25660930914?pr=628#step:5:5)d4ce] — no versions left
   └─ClimaLand [08f4d4ce] log:
     ├─possible versions are: 0.12.2 or uninstalled
     └─ClimaLand [08f4d4ce] is fixed to version 0.12.2
 ```
But if I run 
```
julia --project= -e 'using Pkg; Pkg.develop(path="$(pwd())"); Pkg.develop(path="$(pwd())/lib/ClimaLandSimulations")'
```
locally, it works just fine

Sbozzolo commented:
Maybe you have to add ClimaDiagnostics to the Project.toml for ClimaLandSimulations?

Sbozzolo commented on src/Diagnostics/standard_diagnostic_frequencies.jl:
```diff
@@ -0,0 +1,264 @@
+    t_start,
+    short_names...,
+)
+"""
+    hourly_mins(short_names...; output_writer, t_start)
+
+
+Return a `ScheduledDiagnostics` that computes the hourly min for the given variable.
+"""
```
```suggestion
"""
    hourly_mins(short_names...; output_writer, t_start)

Return a `ScheduledDiagnostics` that computes the hourly min for the given variable.
"""

```

Sbozzolo replied:
(Remove one extra line from the docstring)

Sbozzolo commented on src/Diagnostics/diagnostic.jl:
```diff
@@ -0,0 +1,100 @@
+            filter(field -> field != :compute!, fieldnames(DiagnosticVariable)),
+        ))"
+    )
+
+
```
```suggestion

```

Sbozzolo commented on src/Diagnostics/bucket_compute_methods.jl:
```diff
@@ -0,0 +1,172 @@
+#compute_soil_temperature!(_, _, _, _, model::T) where {T} =
+#    error_diagnostic_variable("soil_temperature", model)
+
+compute_albedo!(_, _, _, _, land_model) =
+    error_diagnostic_variable("albedo", land_model)
```
We can try coming up with a macro to write this

Sbozzolo replied:
Might look like:

```julia
macro generate_error_functions(function_names...)
    functions = Expr[]
    for function_name in function_names
        function_name_sym = Symbol("compute_", function_name, "!")
        body = quote
            function $function_name_sym(_, _, _, _, land_model)
                error_diagnostic_variable($(Symbol(function_name)), land_model)
            end
        end
        push!(functions, body)
    end
    return quote
        $(functions...)
    end
end

@generate_error_functions "net_radiation" "surface_temperature" "surface_specific_humidity"
```

Sbozzolo commented on src/Diagnostics/default_diagnostics.jl:
```diff
@@ -0,0 +1,63 @@
+# Bucket model
+
+"""
+    produce_common_diagnostic_function(period, reduction)
```
```suggestion
    common_diagnostics(period, reduction)
```

Sbozzolo commented on docs/src/diagnostics/developpers_diagnostics.md:
developer is spelled with only one `p`

Sbozzolo commented on docs/src/diagnostics/users_diagnostics.md:
```diff
@@ -0,0 +1,58 @@
+# default diagnostics 
```
This page has to include instruction how to add a diagnostic that is not default

AlexisRenchon commented:
@kmdeck Could you just look at `src/Diagnostics/define_diagnostics.jl` and check if the names, units, and comments look good? No need to review the full PR! 

Thank you!!

AlexisRenchon commented:
@Sbozzolo 

- I moved the experiment to test (there is nothing new in experiment now). The idea is having a global bucket model in test that then calls some Diagnostic functions and test them. To do that, I had to add SciMLBase and ClimaTimeSteppers to the test Project.toml, I am not sure if this is something we are trying to avoid. 
- I added comments and changed some names to `src/Diagnostics/define_diagnostics.jl`
- I added your macro to generate error functions 
- Docs: available_diagnostics works well
- Docs: I added more content to developers and users documentation page, but there is still more to add (and add more links)
- All CI checks passed

Sbozzolo commented on Artifacts.toml:
```diff
@@ -1,9 +1,12 @@
+[era5_land_forcing_data2021]
+git-tree-sha1 = "ec424296df6b60cfe273ac8f981701fbbed0bd8a"
+
+[land_albedo]
+git-tree-sha1 = "5f7141a13994d5b334f1240e1bd5474ad44d96c4"
```
This should probably not be here

Sbozzolo commented on docs/src/diagnostics/make_diagnostic_table.jl:
```diff
@@ -0,0 +1,27 @@
+using PrettyTables
+
+# Print all available diagnostics to an ASCII table
+
```
```suggestion
CL.define_diagnostics!(nothing)
```

AlexisRenchon replied:
I added that, but while this works, we can't use `get_diagnostic_variable` when just using ClimaLand.Diagnostics since we don't call `define_diagnostics` in src

Sbozzolo commented on src/Diagnostics/diagnostic.jl:
```diff
@@ -0,0 +1,99 @@
+Return a `DiagnosticVariable` from its `short_name`, if it exists.
+"""
+function get_diagnostic_variable(short_name)
+    haskey(ALL_DIAGNOSTICS, short_name) ||
```
For your reference, as discussed
```julia
@test_throws ErrorException get_diagnostic_variable("bob")
```

AlexisRenchon replied:
I added that test, BUT, currently we do not call `define_diagnostics!` inside src (only in docs), so `get_diagnostic_variable` would throw an error with any variable...

Sbozzolo replied:
You can call define_diagnostics in your test

Sbozzolo commented on src/Diagnostics/diagnostic.jl:
```diff
@@ -0,0 +1,99 @@
+"""
+
+    get_diagnostic_variable!(short_name)
+
+
```
```suggestion

```

Sbozzolo commented on src/Diagnostics/bucket_compute_methods.jl:
```diff
@@ -0,0 +1,189 @@
+end
+
+@generate_error_functions "albedo" "net_radiation" "surface_temperature" "surface_specific_humidity" "latent_heat_flux" "aerodynamic_resistance" "sensible_heat_flux" "vapor_flux" "surface_air_density" "soil_temperature" "subsurface_water_storage" "surface_water_content" "snow_water_equivalent"
+
```
To test this
```julia
@test_throws ErrorException compute_albedo!(_, _, _, _, Nothing)
```

AlexisRenchon replied:
```
julia> using Test

julia> using ClimaLand

julia> @test isdefined(ClimaLand.Diagnostics, :compute_albedo!)
Test Passed

julia> @test_throws ErrorException("cannot albedo with nothing") ClimaLand.Diagnostics.comput
e_albedo!(_, _, _, _, nothing)
ERROR: syntax: all-underscore identifier used as rvalue around REPL[9]:1
Stacktrace:
 [1] top-level scope
   @ REPL[9]:1
```

Sbozzolo commented on docs/src/diagnostics/users_diagnostics.md:
````diff
@@ -0,0 +1,133 @@
+```
+
+Your diagnostics have now been written in netcdf files in your output folder. 
+
````
What if I want to compute some existing diagnostics but with different Writers? Or with different frequencies?

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+        long_name = "Net Radiation",
+        standard_name = "net_radiation",
+        units = "W m^-2",
+        comments = "Balance between incoming and outgoing shortwave and longwave radiation at the land surface.",
```
"Difference" instead of balance?. Maybe we can also note that a positive value means radiation is absorbed by land?

AlexisRenchon replied:
I am not sure where, but we should say somewhere that our sign convention throughout is relative to atmosphere or land, e.g., positive if atmosphere gain, negative if atmosphere loss... (is our sign actually consistent for all? e.g., negative values for rain?)

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+        long_name = "Surface Temperature",
+        standard_name = "surface_temperature",
+        units = "K",
+        comments = "Temperature at the land-atmosphere interface.",
```
How is T_sfc defined? In the bucket, I think we use the value at the top center. that might be better described as "Temperature of the land surface". But if this is defined as a skin temperature (or something), what you have is OK.

AlexisRenchon replied:
I will leave this comment up for now. We should decide soon. What do you think @braghiere ?

braghiere replied:
I agree with @kmdeck. Usually surface temperature refers to skin temperature (a combination of soil + canopy). Temperature at the land-atmosphere interface sounds more like skin temperature to me. How is this term calculated? What does this actually physically represent?

AlexisRenchon replied:
I believe this is the temperature at the center of the top layer, as Kat explained, soil layers being discretized to n layers, n defined by users. We have T_sfc stored for the bucket, but no T_sfc for soilcanopymodel, at the moment, I believe (maybe this has changed). We could rename T_sfc to T_skin. (we have T at all soil layers for both models, but T_sfc only for bucket)

We should meet, the 3 of us, to discuss ClimaLand diagnostic variables. 
Sometimes, we compute the same things across models (bucket, soilcanopymodel, other), they will be stored differently (e.g., p.soil.T or Y.bucket.T), but will have the same name, units, comments.

Before we merge this PR, it will be good to go through the list of variables, and make sure we agree on all the naming (short long standard), units, and comments. I will have something for all, but we should make sure we all agree and it looks good together.

braghiere replied:
Sounds good to me.

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+        long_name = "Latent Heat Flux",
+        standard_name = "latent_heat_flux",
+        units = "W m^-2",
+        comments = "Exchange of energy at the land-atmosphere interface due to water evaporation.",
```
maybe "due to evaporation/sublimation"

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+    # Vapor flux
+    add_diagnostic_variable!(
+        short_name = "vflux",
+        long_name = "Liquid water evaporation",
```
"Water vapor flux"? since this could be sublimation too

kmdeck replied:
also as a comment - this is converted from a mass flux into an equivalent volume flux of liquid water. maybe this is what you meant

AlexisRenchon replied:
`"Flux of water from the land surface to the atmosphere. E.g., evaporation or sublimation."`
Not sure if I should be e.g., or i.e., I guess there's no transpiration in bucket, so i.e., may be ok (if these are the only 2 options)

the units are `"m s^-1"`, correct? I guess this implies this is a volume of water so no need to specify in comments?

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+        long_name = "Surface Air Density",
+        standard_name = "surface_air_density",
+        units = "kg m^−3",
+        comments = "Density of air and the land-atmosphere interface.",
```
and -> at

kmdeck commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,156 @@
+        long_name = "subsurface Water Storage",
+        standard_name = "subsurface_water_storage",
+        units = "m",
+        comments = "Soil water content at multiple depth.",
```
W here is a scalar, not at each layer

kmdeck reviewed (commented):
Looks good in src/Diagnostics/define_diagnostics.jl! Thanks!

kmdeck commented on src/Diagnostics/Diagnostics.jl:
```diff
@@ -4,8 +4,7 @@ import ClimaComms
 
-using ClimaLand
-import ClimaLand: SoilCanopyModel  
+import ..SoilCanopyModel
```
ah ha!

Sbozzolo replied:
The problem was that `Diagnostics.jl`  was imported before the soil module

Sbozzolo commented on .buildkite/pipeline.yml:
```diff
@@ -141,7 +141,7 @@ steps:
         key: "global_bucket_temporalmap_cpu"
         command: "julia --color=yes --project=.buildkite experiments/standalone/Bucket/global_bucket_temporalmap.jl"
-        artifact_paths: "experiments/standalone/Bucket/artifacts_temporalmap/*cpu*"
+        artifact_paths: "experiments/standalone/Bucket/artifacts_temporalmap/*cpu*" 
```
```suggestion
        artifact_paths: "experiments/standalone/Bucket/artifacts_temporalmap/*cpu*"
```

Sbozzolo commented on src/Diagnostics/default_diagnostics.jl:
```diff
@@ -0,0 +1,113 @@
+    default_outputs =
+        hourly_averages(soilcanopy_diagnostics...; output_writer, t_start)
+    return [default_outputs...]
+end
```
Add new line here

Sbozzolo commented on src/Diagnostics/define_diagnostics.jl:
```diff
@@ -0,0 +1,588 @@
+        long_name = "Soil Aerodynamic Resistance",
+        standard_name = "soil_aerodynamic_resistance",
+        units = "",
+        comments = "Soil aerodynamic resistance.",
```
If comments are useless, let's not put them, or let's write `TODO` instead.

Sbozzolo commented on src/Diagnostics/soilcanopy_compute_methods.jl:
```diff
@@ -0,0 +1,417 @@
+    if isnothing(out)
+        return copy(p.soil.R_n)
+    else
+        out .= p.R_n
```
```suggestion
        out .= p.soil.R_n
```

Sbozzolo commented on src/Diagnostics/soilcanopy_compute_methods.jl:
```diff
@@ -0,0 +1,417 @@
+    if isnothing(out)
+        return copy(p.bucket.turbulent_fluxes.shf)
+    else
+        out .= p.bucket.turbulent_fluxes.shf
```
Is this `p.bucket`?

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+
+
+# ClimaDiagnostics
+output_dir = "output/"
```
Let's use a more descriptive name, maybe `global_bucket_function`? 

This will help when we add the diagnostics to more experiments

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+
+nc_writer = ClimaDiagnostics.Writers.NetCDFWriter(space, output_dir)
+
+diags =
+    ClimaLand.CLD.default_diagnostics(model, 1.0; output_writer = nc_writer)
```
This 1.0 is unclear

AlexisRenchon replied:
I agree. I just put a value for `t_start`, but maybe we don't need it? It's an optional arg in `EveryDtSchedule(period; t_start)` called in `common_diagnostics`

Sbozzolo replied:
Yes, `t_start` is good!

AlexisRenchon replied:
Let's come back to this too - I am not sure I followed, do we keep the 1.0? Or adjust something?

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+
+#### ClimaAnalysis ####
+
+simdir = ClimaAnalysis.SimDir("output")
```
```suggestion
simdir = ClimaAnalysis.SimDir(output_dir)
```

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+alpha.dims
+fig = CairoMakie.Figure(size = (400, 600))
+viz.plot!(fig, alpha, time = 2 * 86400)
+CairoMakie.save("alpha.png", fig)
```
```suggestion
CairoMakie.save(joinpath(output_dir, "alpha.png"), fig)
```

AlexisRenchon commented on src/Diagnostics/bucket_compute_methods.jl:
```diff
@@ -0,0 +1,189 @@
+error_diagnostic_variable(variable, land_model::T) where {T} =
+    error("Cannot compute $variable with model = $T")
+
+macro generate_error_functions(function_names...)
```
the test didn't work, let's discuss this when we meet

AlexisRenchon commented on .github/workflows/ClimaLandSimulations.yml:
```diff
@@ -19,13 +19,13 @@ jobs:
         with:
           version: '1.10'
-      - uses: julia-actions/cache@v2
+          #    - uses: julia-actions/cache@v2
```
Let's go back to this at some point, as discussed, this still doesn't work

Sbozzolo replied:
Maybe leave a comment on why this is commented out

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,37 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+
+
+# ClimaDiagnostics
+output_dir = "global_bucket_function/"
+ClimaUtilities.OutputPathGenerator.generate_output_path(output_dir)
```
```suggestion
base_output_dir = "global_bucket_function/"
output_dir = ClimaUtilities.OutputPathGenerator.generate_output_path(base_output_dir)
```

Sbozzolo commented on docs/make.jl:
```diff
@@ -7,7 +7,7 @@ using Distributed
 include("pages_helper.jl")
 tutorials = [
-    "For model developers" => [
+#=    "For model developers" => [
```
Put this back

Sbozzolo commented on docs/src/diagnostics/users_diagnostics.md:
````diff
@@ -0,0 +1,133 @@
+1. define an output folder
+
+```
+mkdir("output")
````
Update this with the OuputPathGenerator

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
 driver_cb = ClimaLand.DriverUpdateCallback(updateat, updatefunc)
 cb = SciMLBase.CallbackSet(driver_cb, saving_cb)
 
+
```
This
`saving_cb = ClimaLand.NonInterpSavingCallback(saved_values, saveat);` is no longer required if we use diagnostics

Sbozzolo replied:
From L161 to L168

AlexisRenchon replied:
Let's talk about this, 
```
W = [
    Array(Remapping.interpolate(remapper, sol.u[k].bucket.W)) for
    k in 1:length(sol.t)
];
```
returns a `9-element Vector{Matrix{Float64}}`

would be equivalent to ( ? )
```
W = get(simdir; short_name = "wsoil")
W = W.data
```
which returns `23×180×90 Array{Float64, 3}`

(do we need to adjust format a bit, or am I getting the wrong this?)

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+simdir = ClimaAnalysis.SimDir(output_dir)
+println(summary(simdir))
+alpha = get(simdir; short_name = "alpha")
+alpha.dims
```
```suggestion
```

Sbozzolo commented on experiments/standalone/Bucket/global_bucket_function.jl:
```diff
@@ -164,6 +169,38 @@ updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+
+simdir = ClimaAnalysis.SimDir(output_dir)
+println(summary(simdir))
+alpha = get(simdir; short_name = "alpha")
+alpha.dims
+fig = CairoMakie.Figure(size = (400, 600))
+viz.plot!(fig, alpha, time = 2 * 86400)
+CairoMakie.save(joinpath(output_dir, "alpha.png"), fig)
```
```suggestion
for short_name in ["alpha", "W"]
var = get(simdir; short_name)
fig = CairoMakie.Figure(size = (400, 600))
viz.plot!(fig, alpha, time = 2 * 86400)
CairoMakie.save(joinpath(output_dir, "$short_name.png"), fig)
end
```

AlexisRenchon replied:
Done and it works, but two issues: 

- tsoil is 3D (has depth) so viz.plot! requires z. But other variables don't have z. For now I just commented out tsoil, but this is an easy fix.
- more importantly, in this experiment it seems all variables are constant in space, so the heatmap just shows one color.

Sbozzolo commented on lib/ClimaLandSimulations/Project.toml:
```diff
@@ -42,9 +43,12 @@ cuDNN = "02a925ec-e4fe-4b08-9a7e-0d78e3d38ccd"
+ClimaComms = "0.5.6, 0.6"
+ClimaCore = "0.13.2, 0.14"
+ClimaDiagnostics = "0.2"
+ClimaParams = "0.10.2"
```
Double check if you need this

AlexisRenchon replied:
I think we should comeback to `ClimaLandSimulations` in another PR, it has a lot of things to adjust...

Sbozzolo commented on src/ClimaLand.jl:
```diff
@@ -318,4 +318,9 @@ include("integrated/soil_energy_hydrology_biogeochemistry.jl")
+include(joinpath("Diagnostics", "Diagnostics.jl"))
+import .Diagnostics as CLD # ClimaLand Diagnostics
+
+
```
```suggestion
```

AlexisRenchon commented:
Let's discuss the Bucket experiment a bit, we want to use an existing experiment, but it would be good if variables vary in space at least... I am not sure it is the case with `global_bucket_function.jl` (our heatmap plots are 1 color), maybe we should use another. We were using `global_bucket_temporalmap.jl` because of this, which has albedo varying in space (read from a file), but it defines the bucket model inside a function that I commented out before... and we want to not change the previous code but add diagnostics to it. Anyway, let's discuss.

Sbozzolo commented:
Let's not forget about the issue on `@generate_error_functions`

Sbozzolo reviewed (approved):
Some work is left for a future PR, but this looks good!

Thank you very much!
